library/src/main/camera2/com/google/android/cameraview/Camera2.java

```diff
@@ -478,15 +585,46 @@ private void startCaptureSession() {
         mSurfaceInfo.surface.setDefaultBufferSize(previewSize.getWidth(), previewSize.getHeight());
         Surface surface = new Surface(mSurfaceInfo.surface);
         try {
+            CameraCharacteristics characteristics = mCameraManager.getCameraCharacteristics(mCameraId);
+            StreamConfigurationMap map = characteristics
+                    .get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);
+            mVideoSize = chooseVideoSize(map.getOutputSizes(MediaRecorder.class), mMinVideoWidth, mMinVideoHeight);
             mPreviewRequestBuilder = mCamera.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
             mPreviewRequestBuilder.addTarget(surface);
-            mCamera.createCaptureSession(Arrays.asList(surface, mImageReader.getSurface()),
-                    mSessionCallback, null);
+            if (mStartVideoRecording) {
+                mPreviewRequestBuilder.addTarget(mMediaRecorder.getSurface());
+                mCamera.createCaptureSession(Arrays.asList(surface, mMediaRecorder.getSurface()),
+                                             mSessionCallback, null);
+            } else {
+                mCamera.createCaptureSession(Arrays.asList(surface, mImageReader.getSurface()),
+                                             mSessionCallback, null);
+            }
         } catch (CameraAccessException e) {
             throw new RuntimeException("Failed to start camera session");
         }
     }
 
+    private android.util.Size chooseVideoSize(android.util.Size[] choices, int width, int height) {
+        List<android.util.Size> bigEnough = new ArrayList<>();
+        for (android.util.Size option : choices) {
+            if (option.getWidth() >= width && option.getHeight() >= height) {
+                bigEnough.add(option);
+            }
+        }
+
+        if (bigEnough.size() > 0) {
+            return Collections.min(bigEnough, new Comparator<android.util.Size>() {
+                @Override
+                public int compare(android.util.Size size1, android.util.Size size2) {
+                    return Integer.compare(size1.getWidth() * size1.getHeight(), size2.getWidth() * size2.getHeight());
+                }
+            });
+        } else {
+            Log.e(TAG, "Couldn't find any suitable video recording size");
+            return choices[0];
```

Comment:
I added this: 
```
if(choices == null || choices.length == 0) {
  throw new RuntimeException("Failed to start video recording session. Camera didn't return any output sizes.");
}
```
I think the camera should return at least one supported output size.